File: go/Assembler/template_engine/engine_normal.go
package template_engine

import (
	"assembler/app"
	"assembler/app/json"
	"assembler/template_common"
	"strings"
)

type EngineNormal struct {
	AppViewPrefix string
}

func NewEngineNormal(prefix string) *EngineNormal {
	return &EngineNormal{AppViewPrefix: prefix}
}

func (e *EngineNormal) SetAppViewPrefix(prefix string) {
	e.AppViewPrefix = prefix
}

func (e *EngineNormal) GetAppViewPrefix() string {
	return e.AppViewPrefix
}

// ReplaceCaseInsensitive replaces the first occurrence of 'from' in 'text' (case-insensitive) with 'to'
func ReplaceCaseInsensitive(text, from, to string) string {
	textLower := strings.ToLower(text)
	fromLower := strings.ToLower(from)
	if idx := strings.Index(textLower, fromLower); idx != -1 {
		end := idx + len(from)
		return text[:idx] + to + text[end:]
	}
	return text
}

// MergeTemplates merges templates by replacing placeholders with corresponding HTML
// This is a hybrid method that processes both slotted templates and simple placeholders
// JSON files with matching names are automatically merged with HTML templates before processing
func (e *EngineNormal) MergeTemplates(appSite, appFile, appView string, templates map[string]struct {
	HTML string
	JSON *string
}, enableJsonProcessing bool) string {
	if len(templates) == 0 {
		return ""
	}
	mainHtml, mainJson := e.GetTemplate(appSite, appFile, templates, appView, e.AppViewPrefix, true)
	if mainHtml == "" {
		return ""
	}
	contentHtml := mainHtml
	if enableJsonProcessing && mainJson != nil {
		contentHtml = MergeTemplateWithJson(contentHtml, *mainJson)
	}
	mergedTemplates := make(map[string]string)
	allJsonValues := make(map[string]string)

	// Add main template JSON values to the global collection if it exists
	if enableJsonProcessing && mainJson != nil {
		jsonObj := app.ParseJsonString(*mainJson)
		for k, v := range jsonObj.Iter() {
			if v.Kind == json.JsonString {
				allJsonValues[k] = v.StrVal
			}
		}
	}

	for k, v := range templates {
		htmlContent := v.HTML
		jsonContent := v.JSON
		if enableJsonProcessing && jsonContent != nil {
			htmlContent = MergeTemplateWithJson(htmlContent, *jsonContent)
			// Parse JSON and collect key-value pairs for allJsonValues
			jsonObj := app.ParseJsonString(*jsonContent)
			for jk, jv := range jsonObj.Iter() {
				if jv.Kind == json.JsonString {
					allJsonValues[jk] = jv.StrVal
				}
			}
		}
		mergedTemplates[k] = htmlContent
	}
	previous := ""
	for {
		previous = contentHtml
		contentHtml = e.MergeTemplateSlots(contentHtml, appSite, appView, mergedTemplates)
		contentHtml = e.ReplaceTemplatePlaceholdersWithJson(contentHtml, appSite, mergedTemplates, allJsonValues, appView)
		if contentHtml == previous {
			break
		}
	}
	return contentHtml
}

// MergeTemplateWithJson merges JSON data into the HTML template
// Advanced merge logic for block and conditional patterns
func MergeTemplateWithJson(template, jsonText string) string {
	// Parse JSON using our JsonConverter
	jsonObj := app.ParseJsonString(jsonText)

	// Advanced merge logic for block and conditional patterns
	result := template

	// Instead of finding all array tags first, directly match JSON array keys to template blocks
	for key, value := range jsonObj.Iter() {
		if value.Kind == json.JsonArrayKind && value.ArrVal != nil {
			dataList := value.ArrVal
			// Try to find a matching template block for this JSON array
			keyNorm := strings.ToLower(key)

			// Look for possible template tags that match this JSON key
			possibleTags := []string{
				key,
				keyNorm,
				strings.TrimSuffix(keyNorm, "s"), // Remove trailing 's'
				keyNorm + "s",                    // Add trailing 's'
			}

			for _, tag := range possibleTags {
				blockStartTag := "{{@" + tag + "}}"
				blockEndTag := "{{/" + tag + "}}"

				if startIdx := findCaseInsensitive(result, blockStartTag); startIdx != -1 {
					searchFrom := startIdx + len(blockStartTag)
					if endIdx := findCaseInsensitive(result[searchFrom:], blockEndTag); endIdx != -1 {
						endIdx = searchFrom + endIdx

						if startIdx < endIdx {
							// Found a valid block - process it
							contentStartIdx := startIdx + len(blockStartTag)
							if contentStartIdx <= endIdx {
								blockContent := result[contentStartIdx:endIdx]
								var mergedBlock strings.Builder

								// Find all conditional blocks in the template block (e.g., {{@Key}}...{{/Key}})
								conditionalKeys := make(map[string]bool)
								condIdx := 0
								for condIdx < len(blockContent) {
									if condStart := findCaseInsensitive(blockContent[condIdx:], "{{@"); condStart != -1 {
										condStart = condIdx + condStart
										if condEnd := strings.Index(blockContent[condStart:], "}}"); condEnd != -1 {
											condEnd = condStart + condEnd
											condKey := strings.TrimSpace(blockContent[condStart+3 : condEnd])
											conditionalKeys[condKey] = true
											condIdx = condEnd + 2
										} else {
											break
										}
									} else {
										break
									}
								}

								for i := 0; i < dataList.Len(); i++ {
									if item, exists := dataList.Get(i); exists && item.Kind == json.JsonObjectKind && item.ObjVal != nil {
										itemObj := item.ObjVal
										itemBlock := blockContent

										// Replace all placeholders dynamically
										for k, v := range itemObj.Iter() {
											placeholder := "{{$" + k + "}}"
											var valueStr string
											switch v.Kind {
											case json.JsonString:
												valueStr = v.StrVal
											case json.JsonNumber:
												valueStr = v.String()
											case json.JsonInteger:
												valueStr = v.String()
											case json.JsonBool:
												valueStr = v.String()
											default:
												valueStr = ""
											}
											itemBlock = replaceAllCaseInsensitive(itemBlock, placeholder, valueStr)
										}

										// Handle all conditional blocks dynamically
										for condKey := range conditionalKeys {
											condValue := false
											// Try case-insensitive lookup for the conditional key
											for objKey, condObj := range itemObj.Iter() {
												if strings.EqualFold(objKey, condKey) {
													switch condObj.Kind {
													case json.JsonBool:
														condValue = condObj.BoolVal
													case json.JsonString:
														// Parse string as bool, or check if non-empty
														if condObj.StrVal == "true" {
															condValue = true
														} else if condObj.StrVal == "false" {
															condValue = false
														} else {
															condValue = condObj.StrVal != ""
														}
													case json.JsonInteger:
														condValue = condObj.IntVal != 0
													case json.JsonNumber:
														condValue = condObj.NumVal != 0.0
													default:
														condValue = false
													}
													break
												}
											}
											itemBlock = handleConditional(itemBlock, condKey, condValue)
										}
										mergedBlock.WriteString(itemBlock)
									}
								}

								// Replace block in result
								result = result[:startIdx] + mergedBlock.String() + result[endIdx+len(blockEndTag):]
								break // Process only the first matching template for this JSON key
							}
						}
					}
				}
			}
		}
	}

	// Handle {{^ArrayName}} block if array is empty (dynamic detection)
	for key, value := range jsonObj.Iter() {
		emptyBlockStart := "{{^" + key + "}}"
		emptyBlockEnd := "{{/" + key + "}}"

		if emptyStartIdx := findCaseInsensitive(result, emptyBlockStart); emptyStartIdx != -1 {
			if emptyEndIdx := findCaseInsensitive(result, emptyBlockEnd); emptyEndIdx != -1 {
				if value.Kind == json.JsonArrayKind && value.ArrVal != nil {
					isEmpty := value.ArrVal.Len() == 0
					emptyContent := result[emptyStartIdx+len(emptyBlockStart) : emptyEndIdx]
					if isEmpty {
						result = result[:emptyStartIdx] + emptyContent + result[emptyEndIdx+len(emptyBlockEnd):]
					} else {
						result = result[:emptyStartIdx] + result[emptyEndIdx+len(emptyBlockEnd):]
					}
				}
			}
		}
	}

	// Replace remaining simple placeholders
	for key, value := range jsonObj.Iter() {
		if value.Kind == json.JsonString {
			placeholder := "{{$" + key + "}}"
			result = replaceAllCaseInsensitive(result, placeholder, value.StrVal)
		}
	}

	return result
}

// GetTemplate retrieves a template (html and json) from the templates dictionary with AppView fallback
func (e *EngineNormal) GetTemplate(appSite, templateName string, templates map[string]struct {
	HTML string
	JSON *string
}, appView, appViewPrefix string, useAppViewFallback bool) (string, *string) {
	if len(templates) == 0 {
		return "", nil
	}
	viewPrefix := appViewPrefix

	// FIRST: Check for AppView-specific template resolution when AppView context is provided
	if useAppViewFallback && appView != "" && viewPrefix != "" {
		// Case-insensitive check if template_name contains view_prefix
		templateNameLower := strings.ToLower(templateName)
		viewPrefixLower := strings.ToLower(viewPrefix)

		if strings.Contains(templateNameLower, viewPrefixLower) {
			// Direct replacement: Replace the AppViewPrefix with the AppView value
			// For example: Html3AContent with AppViewPrefix=Html3A and AppView=html3B becomes html3BContent
			appKey := ReplaceCaseInsensitive(templateName, viewPrefix, appView)
			fallbackTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(appKey)
			for k := range templates {
				if strings.EqualFold(k, fallbackTemplateKey) {
					v := templates[k]
					return v.HTML, v.JSON // Found AppView-specific template, use it
				}
			}
		}
	}

	// SECOND: If no AppView-specific template found, try primary template
	primaryTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(templateName)
	for k := range templates {
		if strings.EqualFold(k, primaryTemplateKey) {
			v := templates[k]
			return v.HTML, v.JSON
		}
	}

	return "", nil
}

// ReplaceTemplatePlaceholdersWithJson replaces placeholders using both templates and JSON values
func (e *EngineNormal) ReplaceTemplatePlaceholdersWithJson(html, appSite string, htmlFiles, jsonValues map[string]string, appView string) string {
	result := html
	searchPos := 0
	for searchPos < len(result) {
		openStart := strings.Index(result[searchPos:], "{{")
		if openStart == -1 {
			break
		}
		openStart += searchPos
		if openStart+2 < len(result) && strings.ContainsAny(string(result[openStart+2]), "#@/") { // Don't skip '$' placeholders!
			searchPos = openStart + 2
			continue
		}
		closeStart := strings.Index(result[openStart+2:], "}}")
		if closeStart == -1 {
			break
		}
		closeStart += openStart + 2
		placeholderName := strings.TrimSpace(result[openStart+2 : closeStart])
		if placeholderName == "" {
			searchPos = openStart + 2
			continue
		}

		var processedReplacement string

		// PRIORITY 1: Check for JSON placeholders first (starts with '$')
		if strings.HasPrefix(placeholderName, "$") {
			key := placeholderName[1:] // Remove the leading '$'
			if value, exists := jsonValues[key]; exists {
				processedReplacement = value
			}
		} else if template_common.IsAlphaNumeric(placeholderName) {
			// PRIORITY 2: Check for template placeholders (alphanumeric)
			templateKey := strings.ToLower(appSite) + "_" + strings.ToLower(placeholderName)

			// Check for AppView fallback FIRST if appView is provided
			if appView != "" && e.AppViewPrefix != "" {
				// Case-insensitive check if placeholderName contains AppViewPrefix
				placeholderLower := strings.ToLower(placeholderName)
				prefixLower := strings.ToLower(e.AppViewPrefix)
				if strings.Contains(placeholderLower, prefixLower) {
					// Replace AppViewPrefix with AppView in placeholderName
					appKey := ReplaceCaseInsensitive(placeholderName, e.AppViewPrefix, appView)
					fallbackTemplateKey := strings.ToLower(appSite) + "_" + strings.ToLower(appKey)
					if fallbackContent, exists := htmlFiles[fallbackTemplateKey]; exists {
						processedReplacement = e.ReplaceTemplatePlaceholdersWithJson(fallbackContent, appSite, htmlFiles, jsonValues, appView)
					}
				}
			}

			// If no AppView fallback found, use primary template key
			if processedReplacement == "" {
				if templateContent, exists := htmlFiles[templateKey]; exists {
					processedReplacement = e.ReplaceTemplatePlaceholdersWithJson(templateContent, appSite, htmlFiles, jsonValues, appView)
				}
			}

			// PRIORITY 3: If no template found, try JSON values (for backward compatibility)
			if processedReplacement == "" {
				if value, exists := jsonValues[placeholderName]; exists {
					processedReplacement = value
				}
			}
		}

		if processedReplacement != "" {
			placeholder := result[openStart : closeStart+2]
			result = strings.Replace(result, placeholder, processedReplacement, 1)
			searchPos = openStart + len(processedReplacement)
		} else {
			searchPos = closeStart + 2
		}
	}
	return result
}

// MergeTemplateSlots recursively merges slotted templates with content
func (e *EngineNormal) MergeTemplateSlots(contentHtml, appSite, appView string, templates map[string]string) string {
	if contentHtml == "" || len(templates) == 0 {
		return contentHtml
	}
	previous := ""
	for {
		previous = contentHtml
		contentHtml = e.ProcessTemplateSlots(contentHtml, appSite, appView, templates)
		if contentHtml == previous {
			break
		}
	}
	return contentHtml
}

// ProcessTemplateSlots processes slotted templates using IndexOf
func (e *EngineNormal) ProcessTemplateSlots(contentHtml, appSite, appView string, templates map[string]string) string {
	result := contentHtml
	searchPos := 0
	for searchPos < len(result) {
		openStart := strings.Index(result[searchPos:], "{{#")
		if openStart == -1 {
			break
		}
		openStart += searchPos
		openEnd := strings.Index(result[openStart+3:], "}}")
		if openEnd == -1 {
			break
		}
		openEnd += openStart + 3
		templateName := strings.TrimSpace(result[openStart+3 : openEnd])
		if templateName == "" || !template_common.IsAlphaNumeric(templateName) {
			searchPos = openStart + 1
			continue
		}
		closeTag := "{{/" + templateName + "}}"
		closeStart, found := template_common.FindMatchingCloseTag(result, openEnd+2, "{{#"+templateName+"}}", closeTag)
		if !found {
			searchPos = openStart + 1
			continue
		}
		innerStart := openEnd + 2
		innerContent := result[innerStart:closeStart]

		// Convert string map to anonymous struct map for GetTemplate method (like Rust does)
		templatesForGetTemplate := make(map[string]struct {
			HTML string
			JSON *string
		})
		for k, v := range templates {
			templatesForGetTemplate[k] = struct {
				HTML string
				JSON *string
			}{HTML: v, JSON: nil}
		}

		// Use GetTemplate method like Rust does
		templateHtml, _ := e.GetTemplate(appSite, templateName, templatesForGetTemplate, appView, e.AppViewPrefix, true)
		if templateHtml != "" {
			slotContents := e.ExtractSlotContents(innerContent, appSite, appView, templates)
			processedTemplate := templateHtml
			for k, v := range slotContents {
				processedTemplate = strings.ReplaceAll(processedTemplate, k, v)
			}
			fullMatch := result[openStart : closeStart+len(closeTag)]
			result = strings.Replace(result, fullMatch, processedTemplate, 1)
			searchPos = openStart + len(processedTemplate)
		} else {
			searchPos = openStart + 1
		}
	}
	return result
}

// ExtractSlotContents extracts slot contents using IndexOf approach
func (e *EngineNormal) ExtractSlotContents(innerContent, appSite, appView string, templates map[string]string) map[string]string {
	slotContents := make(map[string]string)
	searchPos := 0
	for searchPos < len(innerContent) {
		slotStart := strings.Index(innerContent[searchPos:], "{{@HTMLPLACEHOLDER")
		if slotStart == -1 {
			break
		}
		slotStart += searchPos
		afterPlaceholder := slotStart + 18
		slotNum := ""
		pos := afterPlaceholder
		for pos < len(innerContent) && innerContent[pos] >= '0' && innerContent[pos] <= '9' {
			slotNum += string(innerContent[pos])
			pos++
		}
		if pos+1 >= len(innerContent) || innerContent[pos:pos+2] != "}}" {
			searchPos = slotStart + 1
			continue
		}
		slotOpenEnd := pos + 2
		var closeTag, openTag string
		if slotNum == "" {
			closeTag = "{{/HTMLPLACEHOLDER}}"
			openTag = "{{@HTMLPLACEHOLDER}}"
		} else {
			closeTag = "{{/HTMLPLACEHOLDER" + slotNum + "}}"
			openTag = "{{@HTMLPLACEHOLDER" + slotNum + "}}"
		}
		closeStart, found := template_common.FindMatchingCloseTag(innerContent, slotOpenEnd, openTag, closeTag)
		if !found {
			searchPos = slotStart + 1
			continue
		}
		slotContent := innerContent[slotOpenEnd:closeStart]
		var slotKey string
		if slotNum == "" {
			slotKey = "{{$HTMLPLACEHOLDER}}"
		} else {
			slotKey = "{{$HTMLPLACEHOLDER" + slotNum + "}}"
		}
		recursiveResult := e.MergeTemplateSlots(slotContent, appSite, appView, templates)
		recursiveResult = e.ReplaceTemplatePlaceholders(recursiveResult, appSite, appView, templates)
		slotContents[slotKey] = recursiveResult
		searchPos = closeStart + len(closeTag)
	}
	return slotContents
}

// ReplaceTemplatePlaceholders processes simple placeholders only
func (e *EngineNormal) ReplaceTemplatePlaceholders(html, appSite, appView string, htmlFiles map[string]string) string {
	result := html
	searchPos := 0
	for searchPos < len(result) {
		openStart := strings.Index(result[searchPos:], "{{")
		if openStart == -1 {
			break
		}
		openStart += searchPos
		if openStart+2 < len(result) && strings.ContainsAny(string(result[openStart+2]), "#@$/") {
			searchPos = openStart + 2
			continue
		}
		closeStart := strings.Index(result[openStart+2:], "}}")
		if closeStart == -1 {
			break
		}
		closeStart += openStart + 2
		placeholderName := strings.TrimSpace(result[openStart+2 : closeStart])
		if placeholderName == "" || !template_common.IsAlphaNumeric(placeholderName) {
			searchPos = openStart + 2
			continue
		}
		templateContent := htmlFiles[placeholderName]
		var processedReplacement string
		if templateContent != "" {
			processedReplacement = e.ReplaceTemplatePlaceholders(templateContent, appSite, appView, htmlFiles)
		}
		if processedReplacement != "" {
			placeholder := result[openStart : closeStart+2]
			result = strings.Replace(result, placeholder, processedReplacement, 1)
			searchPos = openStart + len(processedReplacement)
		} else {
			searchPos = closeStart + 2
		}
	}
	return result
}

// Helper: Replace all case-insensitive occurrences
func replaceAllCaseInsensitive(input, search, replacement string) string {
	result := input
	idx := 0
	for {
		if found := findCaseInsensitive(result[idx:], search); found != -1 {
			found = idx + found
			result = result[:found] + replacement + result[found+len(search):]
			idx = found + len(replacement)
		} else {
			break
		}
	}
	return result
}

// Helper: Handle conditional blocks like {{@Selected}}...{{/Selected}}
func handleConditional(input, key string, condition bool) string {
	result := input

	// Support spaces inside block tags, e.g. {{@Selected}} ... {{ /Selected}}
	condStart := "{{@" + key + "}}"
	condEndSpace := "{{ /" + key + "}}"

	// Handle first pattern: {{ /Key}} (with space)
	for {
		if startIdx := findCaseInsensitive(result, condStart); startIdx != -1 {
			if endIdx := findCaseInsensitive(result, condEndSpace); endIdx != -1 {
				content := result[startIdx+len(condStart) : endIdx]
				if condition {
					result = result[:startIdx] + content + result[endIdx+len(condEndSpace):]
				} else {
					result = result[:startIdx] + result[endIdx+len(condEndSpace):]
				}
			} else {
				break
			}
		} else {
			break
		}
	}

	// Also handle without space: {{/Key}}
	condEndNoSpace := "{{/" + key + "}}"
	for {
		if startIdx := findCaseInsensitive(result, condStart); startIdx != -1 {
			if endIdx := findCaseInsensitive(result, condEndNoSpace); endIdx != -1 {
				content := result[startIdx+len(condStart) : endIdx]
				if condition {
					result = result[:startIdx] + content + result[endIdx+len(condEndNoSpace):]
				} else {
					result = result[:startIdx] + result[endIdx+len(condEndNoSpace):]
				}
			} else {
				break
			}
		} else {
			break
		}
	}

	return result
}

// Helper: Find case-insensitive occurrence of search string
func findCaseInsensitive(haystack, needle string) int {
	lower := strings.ToLower(haystack)
	lowerNeedle := strings.ToLower(needle)
	return strings.Index(lower, lowerNeedle)
}

File: rust/Assembler/src/engine/normal/engine_normal.rs
use crate::common::common_util::CommonUtil;
use crate::interface::ILoaderNormal;
use arshu::common::Logger;
use std::collections::HashMap;

/// IndexOf-based template engine implementation for improved performance
/// Matches C# EngineNormal structure - uses ILoaderNormal for template access
pub struct EngineNormal {
    app_view_prefix: String,
}

impl EngineNormal {
    pub fn new(prefix: String) -> Self {
        Self {
            app_view_prefix: prefix,
        }
    }

    pub fn app_view_prefix(&self) -> &str {
        &self.app_view_prefix
    }

    pub fn set_app_view_prefix(&mut self, prefix: String) {
        self.app_view_prefix = prefix;
    }

    /// Merges templates by replacing placeholders with corresponding HTML
    /// This is a hybrid method that processes both slotted templates and simple placeholders
    /// JSON files with matching names are automatically merged with HTML templates before processing
    pub fn merge_templates(
        &self,
        app_site: &str,
        app_file: &str,
        app_view: Option<&str>,
        loader: &dyn ILoaderNormal,
        enable_json_processing: bool,
    ) -> String {
        let app_view_str = app_view.unwrap_or("null");
        Logger::debug(
            &format!(
                "MergeTemplates called: appSite={}, appFile={}, appView={}, enableJson={}",
                app_site, app_file, app_view_str, enable_json_processing
            ),
            Some("EngineNormal"),
        );

        // Get main template using loader (includes AppView fallback and SearchAppSites logic)
        let mut content_html = match loader.get_template_html(
            app_site,
            app_file,
            app_view,
            Some(&self.app_view_prefix),
        ) {
            Some(html) => html,
            None => {
                Logger::warn(
                    &format!(
                        "Main template not found for appSite={}, appFile={}",
                        app_site, app_file
                    ),
                    Some("EngineNormal"),
                );
                return String::new();
            }
        };

        Logger::debug(
            &format!("Main template found, html size: {}", content_html.len()),
            Some("EngineNormal"),
        );

        // Merge main template with JSON using loader's centralized method
        if enable_json_processing {
            Logger::debug("Merging main template with JSON", Some("EngineNormal"));
            content_html = loader.merge_html_with_json(&content_html, app_site, app_file);
            Logger::debug(
                &format!("After main JSON merge: {} chars", content_html.len()),
                Some("EngineNormal"),
            );
        }

        // Simple loop - templates are now loaded on-demand via loader
        let max_passes = 10;
        let mut actual_passes = 0;
        for pass in 0..max_passes {
            let previous = content_html.clone();
            actual_passes = pass + 1;

            Logger::debug(
                &format!("Pass {}, current size: {}", actual_passes, content_html.len()),
                Some("EngineNormal"),
            );

            content_html = self.merge_template_slots(&content_html, app_site, app_view, loader, enable_json_processing);
            Logger::debug(
                &format!("After slot merge: {} chars", content_html.len()),
                Some("EngineNormal"),
            );

            content_html = self.replace_template_placeholders(&content_html, app_site, app_view, loader, enable_json_processing);
            Logger::debug(
                &format!("After placeholder replacement: {} chars", content_html.len()),
                Some("EngineNormal"),
            );

            if content_html == previous {
                Logger::debug(
                    &format!("No changes in pass {}, stopping", actual_passes),
                    Some("EngineNormal"),
                );
                break;
            }
        }

        Logger::debug(
            &format!(
                "MergeTemplates complete after {} passes: output size={}",
                actual_passes,
                content_html.len()
            ),
            Some("EngineNormal"),
        );
        content_html
    }

    /// Gets a template with on-demand loading and JSON merging from ILoaderNormal
    fn get_template_with_json(
        &self,
        app_site: &str,
        template_name: &str,
        loader: &dyn ILoaderNormal,
        app_view: Option<&str>,
        enable_json_processing: bool,
    ) -> Option<String> {
        // Get HTML template (includes AppView fallback and SearchAppSites logic)
        let mut html = loader.get_template_html(app_site, template_name, app_view, Some(&self.app_view_prefix))?;

        Logger::debug(
            &format!(
                "GetTemplateWithJson: template={}, html size={}",
                template_name,
                html.len()
            ),
            Some("EngineNormal"),
        );

        // Merge with JSON if enabled using loader's centralized method
        if enable_json_processing {
            let original_size = html.len();
            html = loader.merge_html_with_json(&html, app_site, template_name);
            Logger::debug(
                &format!(
                    "After JSON merge for {}: size {} -> {}",
                    template_name, original_size, html.len()
                ),
                Some("EngineNormal"),
            );
        }

        Some(html)
    }

    /// Recursively merges slotted templates
    /// Slot patterns: {{#TemplateName}} ... {{@HTMLPLACEHOLDER[N]}} ... {{/HTMLPLACEHOLDER[N]}} ... {{/TemplateName}}
    fn merge_template_slots(
        &self,
        content_html: &str,
        app_site: &str,
        app_view: Option<&str>,
        loader: &dyn ILoaderNormal,
        enable_json_processing: bool,
    ) -> String {
        if content_html.is_empty() {
            return content_html.to_string();
        }

        let mut result = String::with_capacity(content_html.len() * 2);
        let mut search_pos = 0;

        while search_pos < content_html.len() {
            // Find opening slot marker {{#TemplateName}}
            let open_start = match content_html[search_pos..].find("{{#") {
                Some(pos) => search_pos + pos,
                None => {
                    result.push_str(&content_html[search_pos..]);
                    break;
                }
            };

            // Add content before slot
            result.push_str(&content_html[search_pos..open_start]);

            let open_end = match content_html[open_start + 3..].find("}}") {
                Some(pos) => open_start + 3 + pos,
                None => {
                    result.push_str(&content_html[open_start..]);
                    break;
                }
            };

            let template_name = content_html[open_start + 3..open_end].trim();
            let closing_tag = format!("{{{{/{}}}}}", template_name);

            let close_start = match CommonUtil::find_matching_close_tag(
                content_html,
                open_end + 2,
                &format!("{{{{#{}}}}}", template_name),
                &closing_tag,
            ) {
                Some(pos) => pos,
                None => {
                    result.push_str(&content_html[open_start..]);
                    search_pos = open_start + 1;
                    continue;
                }
            };

            // Extract slot content between opening and closing tags (inner content)
            let inner_content = &content_html[open_end + 2..close_start];

            // Get the slot template
            let slot_template = match self.get_template_with_json(
                app_site,
                template_name,
                loader,
                app_view,
                enable_json_processing,
            ) {
                Some(tmpl) => tmpl,
                None => {
                    Logger::warn(
                        &format!("Slot template '{}' not found", template_name),
                        Some("EngineNormal"),
                    );
                    result.push_str(&content_html[open_start..close_start + closing_tag.len()]);
                    search_pos = close_start + closing_tag.len();
                    continue;
                }
            };

            // Extract slot contents from inner_content
            let slot_contents = self.extract_slot_contents(inner_content, app_site, app_view, loader, enable_json_processing);

            // Replace slot placeholders in the slotted template
            let mut merged_slot = slot_template.clone();
            for (slot_key, slot_content) in slot_contents {
                merged_slot = merged_slot.replace(&slot_key, &slot_content);
            }

            // Remove any remaining slot placeholders that weren't replaced
            merged_slot = CommonUtil::remove_remaining_slot_placeholders(&merged_slot);

            result.push_str(&merged_slot);
            search_pos = close_start + closing_tag.len();
        }

        result
    }

    /// Extracts slot contents from inner content between {{#SlotName}} and {{/SlotName}}
    /// Returns a HashMap of slot keys ({{$HTMLPLACEHOLDER}}, {{$HTMLPLACEHOLDER1}}, etc.) to their content
    fn extract_slot_contents(
        &self,
        inner_content: &str,
        app_site: &str,
        app_view: Option<&str>,
        loader: &dyn ILoaderNormal,
        enable_json_processing: bool,
    ) -> HashMap<String, String> {
        let mut slot_contents = HashMap::new();
        let mut search_pos = 0;

        while search_pos < inner_content.len() {
            // Look for slot start {{@HTMLPLACEHOLDER
            let slot_start = match inner_content[search_pos..].find("{{@HTMLPLACEHOLDER") {
                Some(pos) => search_pos + pos,
                None => break,
            };

            // Find the number (if any) and closing }}
            let after_placeholder = slot_start + 18; // Length of "{{@HTMLPLACEHOLDER"
            let mut slot_num = String::new();
            let mut pos = after_placeholder;

            // Extract slot number using byte positions
            while pos < inner_content.len() {
                let byte = inner_content.as_bytes()[pos];
                if byte.is_ascii_digit() {
                    slot_num.push(byte as char);
                    pos += 1;
                } else {
                    break;
                }
            }

            // Check for closing }}
            if pos + 1 >= inner_content.len() || &inner_content[pos..pos + 2] != "}}" {
                search_pos = slot_start + 1;
                continue;
            }

            let slot_open_end = pos + 2;

            // Find matching closing tag
            let close_tag = if slot_num.is_empty() {
                "{{/HTMLPLACEHOLDER}}".to_string()
            } else {
                format!("{{{{/HTMLPLACEHOLDER{}}}}}", slot_num)
            };
            let open_tag = if slot_num.is_empty() {
                "{{@HTMLPLACEHOLDER}}".to_string()
            } else {
                format!("{{{{@HTMLPLACEHOLDER{}}}}}", slot_num)
            };

            let close_start = match CommonUtil::find_matching_close_tag(
                inner_content,
                slot_open_end,
                &open_tag,
                &close_tag,
            ) {
                Some(pos) => pos,
                None => {
                    search_pos = slot_start + 1;
                    continue;
                }
            };

            // Extract slot content
            let slot_content = &inner_content[slot_open_end..close_start];

            // Generate slot key
            let slot_key = if slot_num.is_empty() {
                "{{$HTMLPLACEHOLDER}}".to_string()
            } else {
                format!("{{{{$HTMLPLACEHOLDER{}}}}}", slot_num)
            };

            // Process both slotted templates AND simple placeholders in slot content
            let recursive_result = self.merge_template_slots(slot_content, app_site, app_view, loader, enable_json_processing);
            let recursive_result = self.replace_template_placeholders(&recursive_result, app_site, app_view, loader, enable_json_processing);
            slot_contents.insert(slot_key, recursive_result);

            search_pos = close_start + close_tag.len();
        }

        slot_contents
    }

    /// Replaces simple template placeholders {{TemplateName}} with template content
    fn replace_template_placeholders(
        &self,
        content_html: &str,
        app_site: &str,
        app_view: Option<&str>,
        loader: &dyn ILoaderNormal,
        enable_json_processing: bool,
    ) -> String {
        if content_html.is_empty() {
            return content_html.to_string();
        }

        let mut result = String::with_capacity(content_html.len() * 2);
        let mut search_pos = 0;

        while search_pos < content_html.len() {
            let open_start = match content_html[search_pos..].find("{{") {
                Some(pos) => search_pos + pos,
                None => {
                    result.push_str(&content_html[search_pos..]);
                    break;
                }
            };

            result.push_str(&content_html[search_pos..open_start]);

            if open_start + 2 >= content_html.len() {
                result.push_str(&content_html[open_start..]);
                break;
            }

            // Skip special markers
            let next_char = content_html.as_bytes()[open_start + 2];
            if next_char == b'#' || next_char == b'@' || next_char == b'$' || next_char == b'/' {
                result.push_str("{{");
                search_pos = open_start + 2;
                continue;
            }

            let close_start = match content_html[open_start + 2..].find("}}") {
                Some(pos) => open_start + 2 + pos,
                None => {
                    result.push_str(&content_html[open_start..]);
                    break;
                }
            };

            let placeholder_name = content_html[open_start + 2..close_start].trim();

            if placeholder_name.is_empty() || !CommonUtil::is_alphanumeric(placeholder_name) {
                result.push_str(&content_html[open_start..close_start + 2]);
                search_pos = close_start + 2;
                continue;
            }

            // Get and merge template
            match self.get_template_with_json(app_site, placeholder_name, loader, app_view, enable_json_processing) {
                Some(template_content) => {
                    result.push_str(&template_content);
                }
                None => {
                    result.push_str(&content_html[open_start..close_start + 2]);
                }
            }

            search_pos = close_start + 2;
        }

        result
    }
}
